[UnifiedPDF] Tracked text selections should be drawn in the page space

aprotyas opened the pull request:
#### 1c0f305b2a21c9cbf2887b9317afd1e5ac0d6ec5
<pre>
[UnifiedPDF] Tracked text selections should be drawn in the page space
<a href="https://bugs.webkit.org/show_bug.cgi?id=268808">https://bugs.webkit.org/show_bug.cgi?id=268808</a>
<a href="https://rdar.apple.com/122371772">rdar://122371772</a>

Reviewed by Simon Fraser.

Currently, there is no visual indication that we keep track of text
selections on mouse drag. This patch addresses the issue by delegating
the painting of PDFSelection elements to PDFKit, as well as laying some
groundwork to meet that end.

Specifically, this patch does the following:
1. Ask PDFKit to draw the current selection through the PDFSelection SPI
   drawForPage:withBox:active:inContext:. Also provided a forward
   declaration for the interface to keep the build happy.
2. Added hooks to inform the PDF plugin when window activity has
   changed. This allows the PDF plugin to be aware of whether or not the
   window it is in is currently active. This knowledge is vended from
   the plugin to PDFKit to ensure we draw secondary colors when we don&apos;t
   have focus.
3. Drive-by fix: remove the unnecessary concept of &quot;committing&quot; a
   selection, as well as the now redundant &quot;isActive&quot; state. There was
   extra motivation to do this since it may be confusing if it lived
   next to the window activity state.

* Source/WebCore/page/Page.h:
* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::windowActivityDidChange):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::windowActivityDidChange):
(WebKit::UnifiedPDFPlugin::paintPDFContent):
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::beginTrackingSelection):
(WebKit::UnifiedPDFPlugin::continueTrackingSelection):
(WebKit::UnifiedPDFPlugin::setCurrentSelection):
(WebKit::UnifiedPDFPlugin::commitCurrentSelection): Deleted.
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::windowActivityDidChange):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::windowActivityDidChange):
(WebKit::WebPage::setActivityState):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/274246@main">https://commits.webkit.org/274246@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/015877baca7b77d4b297fc269b72a0e18fe9d8d8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38384 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17327 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40721 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40941 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/34127 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/40693 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20100 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14672 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32346 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38957 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14627 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33453 "Failed to checkout and rebase branch from PR 23907") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12721 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12701 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34301 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42220 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/34919 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34713 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/38541 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/13291 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/10999 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36742 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14857 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8633 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/13765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14323 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -290,6 +289,8 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
     SelectionTrackingData m_selectionTrackingData;
     RetainPtr<PDFSelection> m_currentSelection;
+
+    bool m_windowIsActive { true };
```
Can we just fetch this on demand, rather than aliasing the state here (which is bug-prune). Does the Page know?

aprotyas replied:
Yes, I think we totally can. I see that `WebPage::activityState()` is a thing, so I'll just consult that when we need.

aprotyas replied:
Instead of storing state, I just have the WebPage notify the plugin that window activity has changed, so that we can issue a repaint request.

Also, we refer to the activity state information on the `Page` object to determine whether to draw active or inactive colors.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1304,16 +1308,8 @@ static IntRect computeMarqueeSelectionRect(const WebCore::IntPoint& point1, cons
-{
-    if (reason == SelectionCommitReason::ReceivedMouseUp && !std::exchange(m_selectionTrackingData.isActive, false))
-        return;
+    updateLayerHierarchy();
```
This is a big hammer to just issue a repaint. Just need to call setNeedsDisplay() somewhere, ideally constraining the repaint to just the old/new selection rect, or at least the touched pages only.

aprotyas replied:
That's fair. I can do two optimizations here:

1. Constrain the repaint to the bounds of the selection we're painting
2. Constrain the repaint to the touched pages only

aprotyas replied:
Information for both should be readily available, let's see

aprotyas replied:
I wasn't able to correctly do the transformation from the page space to the content layer space... deferring that work for a not-so-distant future optimization since I don't want this to block us.

https://bugs.webkit.org/show_bug.cgi?id=268927

aprotyas replied:
Let me know if you think this may still be too big a hammer for repaint requests.

aprotyas committed:
[UnifiedPDF] Tracked text selections should be drawn in the page space
https://bugs.webkit.org/show_bug.cgi?id=268808
rdar://122371772

Reviewed by Simon Fraser.

Currently, there is no visual indication that we keep track of text
selections on mouse drag. This patch addresses the issue by delegating
the painting of PDFSelection elements to PDFKit, as well as laying some
groundwork to meet that end.

Specifically, this patch does the following:
1. Ask PDFKit to draw the current selection through the PDFSelection SPI
   drawForPage:withBox:active:inContext:. Also provided a forward
   declaration for the interface to keep the build happy.
2. Added hooks to inform the PDF plugin when window activity has
   changed. This allows the PDF plugin to be aware of whether or not the
   window it is in is currently active. This knowledge is vended from
   the plugin to PDFKit to ensure we draw secondary colors when we don't
   have focus.
3. Drive-by fix: remove the unnecessary concept of "committing" a
   selection, as well as the now redundant "isActive" state. There was
   extra motivation to do this since it may be confusing if it lived
   next to the window activity state.

* Source/WebCore/page/Page.h:
* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::windowActivityDidChange):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::windowActivityDidChange):
(WebKit::UnifiedPDFPlugin::paintPDFContent):
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::beginTrackingSelection):
(WebKit::UnifiedPDFPlugin::continueTrackingSelection):
(WebKit::UnifiedPDFPlugin::setCurrentSelection):
(WebKit::UnifiedPDFPlugin::commitCurrentSelection): Deleted.
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::windowActivityDidChange):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::windowActivityDidChange):
(WebKit::WebPage::setActivityState):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: https://commits.webkit.org/274246@main
